api/service.go

```diff
@@ -0,0 +1,408 @@
+	r, err := s.client.Do("GET", "api/mempool/sstx/fees", "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &ticketFeeRate)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (s *Service) GetNHighestTicketFeeRate(nHighest int) (ticketFeeRate *apiTypes.MempoolTicketFees, err error) {
+	r, err := s.client.Do("GET", "api/mempool/sstx/fees/"+strconv.Itoa(nHighest), "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &ticketFeeRate)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (s *Service) GetTicketDetails() (ticketDetails *apiTypes.MempoolTicketDetails, err error) {
+	r, err := s.client.Do("GET", "api/mempool/sstx/details", "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &ticketDetails)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (s *Service) GetNHighestTicketDetails(nHighest int) (ticketDetails *apiTypes.MempoolTicketDetails, err error) {
+	r, err := s.client.Do("GET", "api/mempool/sstx/details/"+strconv.Itoa(nHighest), "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &ticketDetails)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// GetAddress returns the balances and transactions of an address.
+// The returned transactions are sorted by block height, newest blocks first.
+func (s *Service) GetAddress(address string) (addressState *AddressState, err error) {
+	err = s.client.setBlockbookURL(s.chainParams.Name)
+	if err != nil {
+		return
+	}
+
+	r, err := s.client.Do("GET", "api/v2/address/"+address, "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &addressState)
+	if err != nil {
+		return
+	}
+	return
+}
+
+//GetXpub Returns balances and transactions of an xpub
+func (s *Service) GetXpub(xPub string) (xPubBalAndTxs *XpubBalAndTxs, err error) {
+	err = s.client.setBlockbookURL(s.chainParams.Name)
+	if err != nil {
+		return
+	}
+
+	r, err := s.client.Do("GET", "api/v2/xpub/"+xPub, "")
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(r, &xPubBalAndTxs)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// GetTicker returns market ticker data for the supported exchanges.
+func (s *Service) GetTicker(exchange Exchange, market string) (ticker *Ticker, err error) {
+	switch exchange {
+	case Binance:
+		if s.chainParams.Name == chaincfg.MainNetParams().Name {
+			s.client.BaseUrl = binanceBaseUrl
+		} else {
+			s.client.BaseUrl = binanceTestnetBaseUrl
+		}
+
+		symbArr := strings.Split(market, "-")
+		if len(symbArr) != 2 {
+			return ticker, errors.New("Invalid symbol format")
+		}
+
+		symb := strings.Join(symbArr[:], "")
+		return s.getBinanceTicker(symb)
+	case Bittrex:
+		if s.chainParams.Name == chaincfg.TestNet3Params().Name {
+			return ticker, errors.New("Bittrex doesn't support testnet")
+		}
+
+		return s.getBittrexTicker(market)
+	case KuCoin:
+		if s.chainParams.Name == chaincfg.MainNetParams().Name {
+			s.client.BaseUrl = kucoinBaseUrl
+		} else {
+			s.client.BaseUrl = KuCoinTestnetBaseUrl
+		}
```

Comment:
checking `chainParams` like that to set proper `BaseUrl` looks kinda messy to me, there gotta be a better way to do this,

and I assume this `Service` is meant to be used concurrently, meaning changing `BaseUrl` without proper go-routine synchronisation patterns is racy (and probably not what you want)